Add zoom smoothing

341101 opened the pull request.

Plonq commented on src/lib.rs:
```diff
@@ -465,24 +477,30 @@ fn pan_orbit_camera(
-                    )
-                });
+        }
+        if let Projection::Orthographic(ref mut p) = *projection {
```
I can see you've moved this part out of the above block, so that it executes when there aren't scroll events (to keep approaching target). This is fine, but I think we should have it follow the same pattern as the others. In other words, put this `if let {} else {}` in section 4, and add a condition that checks if `pan_orbit.target_radius != radius`. You might want to put `has_moved = true` in the above `else if scroll.abs() > 0.0` block, but it might not be necessary (has_moved may not be necessary at all come to think of it but that's another topic).

I think following the same pattern makes it easier to understand overall.

341101 replied:
I've rearranged the code to make it more readable, and now `has_moved = true;` will be executed when `pan_orbit.target_radius != radius.`

Plonq commented on src/lib.rs:
```diff
@@ -465,24 +477,30 @@ fn pan_orbit_camera(
+        }
+        if let Projection::Orthographic(ref mut p) = *projection {
+            let target = p.scale.lerp(&pan_orbit.target_radius, &(1.0 - pan_orbit.zoom_smoothness));
+            if (target - p.scale).abs() > 0.001 {
```
You are only moving towards the target if the distance is greater. This means technically you'll never actually reach the target, which I think is not ideal.

Let's follow the pattern of alpha/beta, and snap to the target if the distance is less than `0.001`, to guarantee we'll always reach the target.

Plonq reviewed (changes requested):
Hi, thanks for the PR! It looks good! There are some small things I'd like to address.

One general issue I want to raise is that on Mac with a touchpad, the default smoothness of 0.8 feels laggy (I'd record a video but it's impossible to demonstrate), which is presumably due to touchpad zoom being smooth already, due to being pixel-based rather than line based. I think perhaps this smoothing should NOT be applied on pixel based scroll events, since they should already be smooth, and it will more likely just make it feel worse. Do you mind updating the PR so the smoothing only applies to line based scrolling?

Also see comments.

341101 commented:
> Hi, thanks for the PR! It looks good! There are some small things I'd like to address.
> 
> One general issue I want to raise is that on Mac with a touchpad, the default smoothness of 0.8 feels laggy (I'd record a video but it's impossible to demonstrate), which is presumably due to touchpad zoom being smooth already, due to being pixel-based rather than line based. I think perhaps this smoothing should NOT be applied on pixel based scroll events, since they should already be smooth, and it will more likely just make it feel worse. Do you mind updating the PR so the smoothing only applies to line based scrolling?
> 
> Also see comments.

Thank you for the feedback. I've made some changes based on your comments and suggestions.

I've split the scroll variable into scroll_line and scroll_pixel to handle the scrolling events separately. This allows us to apply smoothing only on the line-based scroll events, while the pixel-based scroll events are directly added to the target.

Plonq commented:
> Thank you for the feedback. I've made some changes based on your comments and suggestions.
> 
> I've split the `scroll` variable into `scroll_line` and `scroll_pixel` to handle the scrolling events separately. This allows us to apply smoothing only on the line-based scroll events, while the pixel-based scroll events are directly added to the target.

Sounds good. It looks like you might be still in the middle of making changes, so request a re-review when it's ready.

341101 commented:
> Hi, thanks for the PR! It looks good! There are some small things I'd like to address.
> 
> One general issue I want to raise is that on Mac with a touchpad, the default smoothness of 0.8 feels laggy (I'd record a video but it's impossible to demonstrate), which is presumably due to touchpad zoom being smooth already, due to being pixel-based rather than line based. I think perhaps this smoothing should NOT be applied on pixel based scroll events, since they should already be smooth, and it will more likely just make it feel worse. Do you mind updating the PR so the smoothing only applies to line based scrolling?
> 
> Also see comments.

Interestingly, the unit I get when scaling with my Win11 system touchpad is line-based. Maybe we need to set a threshold for scrolling, if it's less than the threshold then no smoothing is applied?

Plonq commented on src/lib.rs:
```diff
@@ -399,11 +418,16 @@ fn pan_orbit_camera(
+                        scroll_line += delta_scroll;
+                    }
+                    MouseScrollUnit::Pixel => {
+                        scroll_pixel += delta_scroll * 0.05;
```
What's the reason for changing the multiplier from 0.005 to 0.05?

341101 replied:
Typed wrong

Plonq commented on src/lib.rs:
```diff
@@ -399,11 +418,16 @@ fn pan_orbit_camera(
+                    }
                 };
-                scroll += ev.y * multiplier * direction * pan_orbit.zoom_sensitivity;
+                // scroll += ev.y * multiplier * direction * pan_orbit.zoom_sensitivity;
```
Remove commented code

Plonq commented on src/lib.rs:
```diff
@@ -461,28 +485,21 @@ fn pan_orbit_camera(
                 let right = transform.rotation * Vec3::X * -pan.x;
                 let up = transform.rotation * Vec3::Y * pan.y;
                 let translation = (right + up) * multiplier;
+                // println!("translation: {:?}", translation);
```
Remove

Plonq commented on src/lib.rs:
```diff
@@ -517,6 +534,25 @@ fn pan_orbit_camera(
         }
 
         // 4 - Update the camera's transform based on current values
+        if let Some(radius) = pan_orbit.radius {
```
Here, you are unwrapping the radius in `if let`, then `map`ing over it, and finally `unwrap`ing it at the end. We can replace all that with a single `map`, because none of this code makes sense if the radius is `None`.

Plonq reviewed (changes requested):
Your latest changes have broken pixel scrolling unfortunately. Do you have a way of testing pixel scrolling? If not, I'd be happy to help with this PR.

341101 commented:
> Your latest changes have broken pixel scrolling unfortunately. Do you have a way of testing pixel scrolling? If not, I'd be happy to help with this PR.

The pixel scroll multiplier was changed from 0.005 to 0.05, which was an input error. I've changed it back.

341101 commented:
> > > Your latest changes have broken pixel scrolling unfortunately. Do you have a way of testing pixel scrolling? If not, I'd be happy to help with this PR.
> > 
> > 
> > The pixel scrolling multiplier has been mistakenly changed from 0.005 to 0.05. I have corrected it back.
> 
> I actually meant it was broken entirely, not just different sensitivity. It is still broken with your latest changes.
> 
> The video below shows me trying to scroll (using pixel units). You can see it only updates when I start orbiting. I think you've misunderstood what `has_moved` is for. It should be set to true if there is any _input_ from the user. And then the camera's position should be updated if `has_moved == true` OR `target_radius != radius`.
> 
>  Screen.Recording.2023-07-31.at.5.10.39.pm.mov

I see now, I added the `pixel based scroll` directly to the `radius`, hence, in the absence of `line based scroll`, `target_radius` would be equal to `radius` and as a result, the condition `if pan_orbit.target_radius != radius` would not be met to correctly set `has_moved`. I have made the correction now, could you please test it again?

Plonq commented on src/lib.rs:
```diff
@@ -517,6 +530,25 @@ fn pan_orbit_camera(
+                if approx_equal(next_radius, pan_orbit.target_radius) {
+                    next_radius = pan_orbit.target_radius
+                }
+                if let Projection::Orthographic(ref mut p) = *projection {
```
You're only applying the zoom to orthographic cameras for line scrolling, which means pixel scrolling is broken for orthographic cameras.

341101 replied:
It's not broken. I have added both the values of line scrolling and pixel scrolling to target_radius.

Plonq commented on src/lib.rs:
```diff
@@ -517,6 +530,25 @@ fn pan_orbit_camera(
 
         // 4 - Update the camera's transform based on current values
+        pan_orbit.radius = pan_orbit.radius.map(|radius| {
+            if pan_orbit.target_radius != radius {
```
I'd like to move some code around so the radius logic follows the same pattern as alpha/beta. Which involves the following:

1. After line 498, set `has_moved = true`
2. Modify line 553 to include radius, i.e. ` if let (Some(alpha), Some(beta), Some(radius)) = (pan_orbit.alpha, pan_orbit.beta, pan_orbit.radius) {`
3. Add `pan_orbit.target_radius != radius` to the existing `if` condition on line 554, and move this block of code into that if block.

Doing this clearly separates getting input from the user, and applying that to the camera. And following the same pattern makes everything easier to understand as a whole.

341101 commented:
I have encountered an issue where my changes are not correctly reflected in the "Files changed" tab of this PR, despite the fact that they are present when directly accessing the file.I have decided to close this PR and create a new one.

Plonq commented on src/lib.rs:
```diff
@@ -464,25 +484,18 @@ fn pan_orbit_camera(
+            let line_delta = -scroll_line * pan_orbit.target_radius * 0.2;
+            let pixel_delta = -scroll_pixel * pan_orbit.target_radius * 0.2;
+            // Add the calculated deltas to the target radius
+            pan_orbit.target_radius = util::apply_zoom_limits(
```
You are only applying zoom limits to the line scrolling, and not the pixel scrolling

341101 replied:
> You are only applying zoom limits to the line scrolling, and not the pixel scrolling

Yes, I have applied `zoom limits` to `pan_orbit.target_radius + line_delta + pixel_delta`, which includes `pixel scrolling`.

Plonq commented:
> I have encountered an issue where my changes are not correctly reflected in the "Files changed" tab of this PR, despite the fact that they are present when directly accessing the file.I have decided to close this PR and create a new one.

No worries. I'd already started a review here, so just apply my comments to your new PR, if they still apply.

Plonq commented:
Continued here: #30
